// @reclaimprotocol/js-sdk/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __commonJS = (cb, mod) => function __require() {
  return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};

// package.json
var require_package = __commonJS({
  "package.json"(exports2, module2) {
    module2.exports = {
      name: "@reclaimprotocol/js-sdk",
      version: "4.15.0",
      description: "Designed to request proofs from the Reclaim protocol and manage the flow of claims and witness interactions.",
      main: "dist/index.js",
      types: "dist/index.d.ts",
      keywords: [
        "reclaim",
        "protocol",
        "blockchain",
        "proof",
        "verification",
        "identity",
        "claims",
        "witness",
        "sdk",
        "javascript",
        "typescript",
        "decentralized",
        "web3"
      ],
      files: [
        "dist"
      ],
      tsup: {
        entry: [
          "src/index.ts"
        ],
        splitting: false,
        sourcemap: true,
        clean: true
      },
      scripts: {
        build: "sh scripts/build.sh",
        prepare: "npm run build",
        release: "release-it",
        test: "jest",
        "test:watch": "jest --watch",
        "test:coverage": "jest --coverage",
        commitlint: "commitlint --edit"
      },
      repository: {
        type: "git",
        url: "https://github.com/reclaimprotocol/reclaim-js-sdk"
      },
      author: "ali <ali@creatoros.co>",
      license: "See License in <https://github.com/reclaimprotocol/.github/blob/main/LICENSE>",
      bugs: {
        url: "https://github.com/reclaimprotocol/reclaim-js-sdk/issues"
      },
      homepage: "https://github.com/reclaimprotocol/reclaim-js-sdk/",
      publishConfig: {
        registry: "https://registry.npmjs.org/",
        access: "public"
      },
      "release-it": {
        git: {
          commitMessage: "chore: release ${version}",
          tagName: "v${version}"
        },
        npm: {
          publish: true,
          tag: "latest"
        },
        github: {
          release: true
        },
        plugins: {
          "@release-it/conventional-changelog": {
            preset: "angular"
          }
        }
      },
      devDependencies: {
        "@commitlint/cli": "^17.7.1",
        "@commitlint/config-conventional": "^17.7.0",
        "@release-it/conventional-changelog": "^10.0.1",
        "@types/jest": "^30.0.0",
        "@types/qs": "^6.9.11",
        "@types/url-parse": "^1.4.11",
        "@types/uuid": "^9.0.7",
        jest: "^30.1.3",
        "jest-environment-jsdom": "^30.1.2",
        qs: "^6.11.2",
        "release-it": "^19.0.4",
        "ts-jest": "^29.4.1",
        tsup: "^8.0.1",
        typescript: "^5.3.3"
      },
      dependencies: {
        canonicalize: "^2.0.0",
        ethers: "^6.9.1",
        "fetch-retry": "^6.0.0",
        "url-parse": "^1.5.10",
        uuid: "^9.0.1"
      },
      overrides: {
        "@conventional-changelog/git-client": "^2.0.0"
      }
    };
  }
});

// src/index.ts
var index_exports = {};
__export(index_exports, {
  ReclaimProofRequest: () => ReclaimProofRequest,
  assertValidProof: () => assertValidProof,
  clearDeviceCache: () => clearDeviceCache,
  getDeviceType: () => getDeviceType,
  getMobileDeviceType: () => getMobileDeviceType,
  isDesktopDevice: () => isDesktopDevice,
  isMobileDevice: () => isMobileDevice,
  transformForOnchain: () => transformForOnchain,
  verifyProof: () => verifyProof
});
module.exports = __toCommonJS(index_exports);

// src/utils/interfaces.ts
var RECLAIM_EXTENSION_ACTIONS = {
  CHECK_EXTENSION: "RECLAIM_EXTENSION_CHECK",
  EXTENSION_RESPONSE: "RECLAIM_EXTENSION_RESPONSE",
  START_VERIFICATION: "RECLAIM_START_VERIFICATION",
  STATUS_UPDATE: "RECLAIM_STATUS_UPDATE"
};

// src/Reclaim.ts
var import_ethers4 = require("ethers");
var import_canonicalize3 = __toESM(require("canonicalize"));

// src/utils/errors.ts
function createErrorClass(name) {
  return class extends Error {
    constructor(message, innerError) {
      const fullMessage = innerError ? `${message || ""} caused by ${innerError.name}: ${innerError.message}` : message;
      super(fullMessage);
      this.innerError = innerError;
      this.name = name;
      if (innerError) {
        this.stack += `
Caused by: ${innerError.stack}`;
      }
    }
  };
}
var TimeoutError = createErrorClass("TimeoutError");
var ProofNotVerifiedError = createErrorClass("ProofNotVerifiedError");
var SessionNotStartedError = createErrorClass("SessionNotStartedError");
var ProviderNotFoundError = createErrorClass("ProviderNotFoundError");
var SignatureGeneratingError = createErrorClass("SignatureGeneratingError");
var SignatureNotFoundError = createErrorClass("SignatureNotFoundError");
var InvalidSignatureError = createErrorClass("InvalidSignatureError");
var UpdateSessionError = createErrorClass("UpdateSessionError");
var InitSessionError = createErrorClass("InitSessionError");
var ProviderFailedError = createErrorClass("ProviderFailedError");
var InvalidParamError = createErrorClass("InvalidParamError");
var ApplicationError = createErrorClass("ApplicationError");
var InitError = createErrorClass("InitError");
var BackendServerError = createErrorClass("BackendServerError");
var GetStatusUrlError = createErrorClass("GetStatusUrlError");
var NoProviderParamsError = createErrorClass("NoProviderParamsError");
var SetParamsError = createErrorClass("SetParamsError");
var SetContextError = createErrorClass("SetContextError");
var SetSignatureError = createErrorClass("SetSignatureError");
var GetAppCallbackUrlError = createErrorClass("GetAppCallbackUrlError");
var StatusUrlError = createErrorClass("StatusUrlError");
var InavlidParametersError = createErrorClass("InavlidParametersError");
var ProofSubmissionFailedError = createErrorClass("ProofSubmissionFailedError");
var ErrorDuringVerificationError = createErrorClass("ErrorDuringVerificationError");
var CallbackUrlRequiredError = createErrorClass("CallbackUrlRequiredError");

// src/utils/logger.ts
var SimpleLogger = class {
  constructor() {
    this.level = "info";
  }
  setLevel(level) {
    this.level = level;
  }
  shouldLog(messageLevel) {
    const levels = ["error", "warn", "info", "silent"];
    return levels.indexOf(this.level) >= levels.indexOf(messageLevel);
  }
  log(level, message, ...args) {
    if (this.shouldLog(level) && this.level !== "silent") {
      const logFunction = this.getLogFunction(level);
      console.log("current level", this.level);
      logFunction(`[${level.toUpperCase()}]`, message, ...args);
    }
  }
  getLogFunction(level) {
    switch (level) {
      case "error":
        return console.error;
      case "warn":
        return console.warn;
      case "info":
        return console.info;
      default:
        return () => {
        };
    }
  }
  info(message, ...args) {
    this.log("info", message, ...args);
  }
  warn(message, ...args) {
    this.log("warn", message, ...args);
  }
  error(message, ...args) {
    this.log("error", message, ...args);
  }
};
var logger = new SimpleLogger();
function setLogLevel(level) {
  logger.setLevel(level);
}
var logger_default = {
  logger,
  setLogLevel
};

// src/utils/helper.ts
var logger2 = logger_default.logger;
function escapeRegExp(string) {
  return string.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}
function replaceAll(str, find, replace) {
  if (find === "") return str;
  return str.replace(new RegExp(escapeRegExp(find), "g"), replace);
}
function scheduleIntervalEndingTask(sessionId, intervals, onFailureCallback, timeout = 1e3 * 60 * 10) {
  setTimeout(() => {
    if (intervals.has(sessionId)) {
      const message = "Interval ended without receiving proofs";
      onFailureCallback(new TimeoutError(message));
      logger2.info(message);
      clearInterval(intervals.get(sessionId));
      intervals.delete(sessionId);
    }
  }, timeout);
}

// src/utils/constants.ts
var BACKEND_BASE_URL = "https://api.reclaimprotocol.org";
function setBackendBaseUrl(url) {
  BACKEND_BASE_URL = url;
}
var constants = {
  // Default callback URL for Reclaim protocol
  get DEFAULT_RECLAIM_CALLBACK_URL() {
    return `${BACKEND_BASE_URL}/api/sdk/callback?callbackId=`;
  },
  // Default error callback URL for Reclaim protocol
  get DEFAULT_RECLAIM_CANCEL_CALLBACK_URL() {
    return `${BACKEND_BASE_URL}/api/sdk/error-callback?callbackId=`;
  },
  // Default status URL for Reclaim sessions
  get DEFAULT_RECLAIM_STATUS_URL() {
    return `${BACKEND_BASE_URL}/api/sdk/session/`;
  },
  get DEFAULT_ATTESTORS_URL() {
    return `${BACKEND_BASE_URL}/api/attestors`;
  },
  // URL for sharing Reclaim templates
  RECLAIM_SHARE_URL: "https://share.reclaimprotocol.org/verifier/?template=",
  // Chrome extension URL for Reclaim Protocol
  CHROME_EXTENSION_URL: "https://chromewebstore.google.com/detail/reclaim-extension/oafieibbbcepkmenknelhmgaoahamdeh",
  // QR Code API base URL
  QR_CODE_API_URL: "https://api.qrserver.com/v1/create-qr-code/"
};

// src/utils/validationUtils.ts
var import_ethers = require("ethers");
var import_canonicalize2 = __toESM(require("canonicalize"));

// src/utils/strings.ts
var import_canonicalize = __toESM(require("canonicalize"));
function canonicalStringify(params) {
  if (!params) {
    return "";
  }
  return (0, import_canonicalize.default)(params) || "";
}

// src/utils/validationUtils.ts
var logger3 = logger_default.logger;
function validateFunctionParams(params, functionName) {
  params.forEach(({ input, paramName, isString }) => {
    if (input == null) {
      logger3.info(`Validation failed: ${paramName} in ${functionName} is null or undefined`);
      throw new InvalidParamError(`${paramName} passed to ${functionName} must not be null or undefined.`);
    }
    if (isString && typeof input !== "string") {
      logger3.info(`Validation failed: ${paramName} in ${functionName} is not a string`);
      throw new InvalidParamError(`${paramName} passed to ${functionName} must be a string.`);
    }
    if (isString && input.trim() === "") {
      logger3.info(`Validation failed: ${paramName} in ${functionName} is an empty string`);
      throw new InvalidParamError(`${paramName} passed to ${functionName} must not be an empty string.`);
    }
  });
}
function validateFunctionParamsWithFn(param, functionName) {
  if (!param.isValid()) {
    logger3.info(`Validation failed: ${param.paramName} in ${functionName} is not valid`);
    throw new InvalidParamError(`${param.paramName} passed to ${functionName} must be valid.`);
  }
}
function validateParameters(parameters) {
  try {
    if (typeof parameters !== "object" || parameters === null) {
      logger3.info(`Parameters validation failed: Provided parameters is not an object`);
      throw new InavlidParametersError(`The provided parameters is not an object`);
    }
    for (const [key, value] of Object.entries(parameters)) {
      if (typeof key !== "string" || typeof value !== "string") {
        logger3.info(`Parameters validation failed: Provided parameters is not an object of key value pairs of string and string`);
        throw new InavlidParametersError(`The provided parameters is not an object of key value pairs of string and string`);
      }
    }
  } catch (e) {
    logger3.info(`Parameters validation failed: ${e.message}`);
    throw new InavlidParametersError(`Invalid parameters passed to validateParameters.`, e);
  }
}
function validateURL(url, functionName) {
  try {
    new URL(url);
  } catch (e) {
    logger3.info(`URL validation failed for ${url} in ${functionName}: ${e.message}`);
    throw new InvalidParamError(`Invalid URL format ${url} passed to ${functionName}.`, e);
  }
}
function validateRedirectionMethod(method, functionName) {
  try {
    if (method === null || method === void 0) {
      return;
    }
    if (method !== "GET" && method !== "POST") {
      throw new Error(`Invalid redirection method ${method} passed to ${functionName}.`);
    }
  } catch (e) {
    logger3.info(`Redirection method validation failed for ${method} in ${functionName}: ${e.message}`);
    throw new InvalidParamError(`Invalid redirection method ${method} passed to ${functionName}.`, e);
  }
}
function validateRedirectionBody(records, functionName) {
  try {
    if (records === null || records === void 0) {
      return;
    }
    if (Array.isArray(records)) {
      for (const record of records) {
        if ("name" in record && record.name && typeof record.name === "string") {
          if ("value" in record && typeof record.value === "string") {
            continue;
          }
        }
        throw new Error("Record in form entries do not have a valid name and/or value");
      }
    } else {
      throw new Error("Redirection body must be an array of objects with name, and value");
    }
  } catch (e) {
    logger3.info(`Redirection body validation failed for ${records} in ${functionName}: ${e.message}`);
    throw new InvalidParamError(`Invalid redirection body ${records} passed to ${functionName}.`, e);
  }
}
function validateSignature(providerId, signature, applicationId, timestamp) {
  try {
    logger3.info(`Starting signature validation for providerId: ${providerId}, applicationId: ${applicationId}, timestamp: ${timestamp}`);
    const message = (0, import_canonicalize2.default)({ providerId, timestamp });
    if (!message) {
      logger3.info("Failed to canonicalize message for signature validation");
      throw new Error("Failed to canonicalize message");
    }
    const messageHash = import_ethers.ethers.keccak256(new TextEncoder().encode(message));
    let appId = import_ethers.ethers.verifyMessage(
      import_ethers.ethers.getBytes(messageHash),
      import_ethers.ethers.hexlify(signature)
    ).toLowerCase();
    if (import_ethers.ethers.getAddress(appId) !== import_ethers.ethers.getAddress(applicationId)) {
      logger3.info(`Signature validation failed: Mismatch between derived appId (${appId}) and provided applicationId (${applicationId})`);
      throw new InvalidSignatureError(`Signature does not match the application id: ${appId}`);
    }
    logger3.info(`Signature validated successfully for applicationId: ${applicationId}`);
  } catch (err) {
    logger3.info(`Signature validation failed: ${err.message}`);
    if (err instanceof InvalidSignatureError) {
      throw err;
    }
    throw new InvalidSignatureError(`Failed to validate signature: ${err.message}`);
  }
}
function validateContext(context) {
  if (context && !("contextAddress" in context)) {
    try {
      validateFunctionParams([
        { input: context, paramName: "context", isString: false }
      ], "validateContext");
      JSON.parse(canonicalStringify(context));
      return;
    } catch (e) {
      logger3.info(`Context validation failed: Provided JSON serializable context is not valid`);
      throw new InvalidParamError(`The provided context is not valid`);
    }
  }
  if (!context.contextAddress) {
    logger3.info(`Context validation failed: Provided context address in context is not valid`);
    throw new InvalidParamError(`The provided context address in context is not valid`);
  }
  if (!context.contextMessage) {
    logger3.info(`Context validation failed: Provided context message in context is not valid`);
    throw new InvalidParamError(`The provided context message in context is not valid`);
  }
  validateFunctionParams([
    { input: context.contextAddress, paramName: "contextAddress", isString: true },
    { input: context.contextMessage, paramName: "contextMessage", isString: true }
  ], "validateContext");
}
function validateModalOptions(modalOptions, functionName, paramPrefix = "") {
  if (modalOptions.title !== void 0) {
    validateFunctionParams([
      { input: modalOptions.title, paramName: `${paramPrefix}title`, isString: true }
    ], functionName);
  }
  if (modalOptions.description !== void 0) {
    validateFunctionParams([
      { input: modalOptions.description, paramName: `${paramPrefix}description`, isString: true }
    ], functionName);
  }
  if (modalOptions.extensionUrl !== void 0) {
    validateURL(modalOptions.extensionUrl, functionName);
    validateFunctionParams([
      { input: modalOptions.extensionUrl, paramName: `${paramPrefix}extensionUrl`, isString: true }
    ], functionName);
  }
  if (modalOptions.darkTheme !== void 0) {
    if (typeof modalOptions.darkTheme !== "boolean") {
      throw new InvalidParamError(`${paramPrefix}darkTheme prop must be a boolean`);
    }
    validateFunctionParams([
      { input: modalOptions.darkTheme, paramName: `${paramPrefix}darkTheme` }
    ], functionName);
  }
  if (modalOptions.modalPopupTimer !== void 0) {
    if (typeof modalOptions.modalPopupTimer !== "number" || modalOptions.modalPopupTimer <= 0 || !Number.isInteger(modalOptions.modalPopupTimer)) {
      throw new InvalidParamError(`${paramPrefix}modalPopupTimer prop must be a valid time in minutes`);
    }
    validateFunctionParams([
      { input: modalOptions.modalPopupTimer, paramName: `${paramPrefix}modalPopupTimer` }
    ], functionName);
  }
  if (modalOptions.showExtensionInstallButton !== void 0) {
    if (typeof modalOptions.showExtensionInstallButton !== "boolean") {
      throw new InvalidParamError(`${paramPrefix}showExtensionInstallButton prop must be a boolean`);
    }
    validateFunctionParams([
      { input: modalOptions.showExtensionInstallButton, paramName: `${paramPrefix}showExtensionInstallButton` }
    ], functionName);
  }
}

// src/utils/fetch.ts
var import_fetch_retry = __toESM(require("fetch-retry"));
var MAX_RETRIES = 3;
var MAX_RETRY_DELAY_MS = 60 * 1e3;
var isHttpResponseStatusRetryable = (statusCode) => {
  return statusCode === 408 || statusCode === 429 || statusCode >= 500;
};
var getRetryDelay = (response) => {
  const retryAfter = response == null ? void 0 : response.headers.get("Retry-After");
  if (!retryAfter) {
    return void 0;
  }
  const trimmed = retryAfter.trim();
  if (/^\d+$/.test(trimmed)) {
    return Math.min(parseInt(trimmed, 10) * 1e3, MAX_RETRY_DELAY_MS);
  }
  const date = new Date(trimmed);
  if (!isNaN(date.getTime())) {
    return Math.min(Math.max(0, date.getTime() - Date.now()), MAX_RETRY_DELAY_MS);
  }
  return void 0;
};
var http = {
  get client() {
    return (0, import_fetch_retry.default)(globalThis.fetch, {
      retries: MAX_RETRIES,
      retryDelay: function(attempt, _, response) {
        const delay = getRetryDelay(response);
        if (delay !== void 0) {
          return delay;
        }
        return Math.pow(2, attempt) * 1e3;
      },
      retryOn: (attempt, error, response) => {
        if (attempt >= MAX_RETRIES) {
          return false;
        }
        if (response && Number.isInteger(response.status)) {
          return isHttpResponseStatusRetryable(response.status);
        }
        return !!error && error.name !== "AbortError";
      }
    });
  }
};

// src/utils/sessionUtils.ts
var logger4 = logger_default.logger;
function initSession(providerId, appId, timestamp, signature, versionNumber) {
  return __async(this, null, function* () {
    logger4.info(`Initializing session for providerId: ${providerId}, appId: ${appId}`);
    try {
      const response = yield http.client(`${BACKEND_BASE_URL}/api/sdk/init/session/`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ providerId, appId, timestamp, signature, versionNumber })
      });
      const res = yield response.json();
      if (!response.ok) {
        logger4.info(`Session initialization failed: ${res.message || "Unknown error"}`);
        throw new InitSessionError(res.message || `Error initializing session with providerId: ${providerId}`);
      }
      return res;
    } catch (err) {
      logger4.info(`Failed to initialize session for providerId: ${providerId}, appId: ${appId}`, err);
      throw err;
    }
  });
}
function updateSession(sessionId, status) {
  return __async(this, null, function* () {
    logger4.info(`Updating session status for sessionId: ${sessionId}, new status: ${status}`);
    validateFunctionParams(
      [{ input: sessionId, paramName: "sessionId", isString: true }],
      "updateSession"
    );
    try {
      const response = yield http.client(`${BACKEND_BASE_URL}/api/sdk/update/session/`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ sessionId, status })
      });
      const res = yield response.json();
      if (!response.ok) {
        const errorMessage = `Error updating session with sessionId: ${sessionId}. Status Code: ${response.status}`;
        logger4.info(errorMessage, res);
        throw new UpdateSessionError(errorMessage);
      }
      logger4.info(`Session status updated successfully for sessionId: ${sessionId}`);
      return res;
    } catch (err) {
      const errorMessage = `Failed to update session with sessionId: ${sessionId}`;
      logger4.info(errorMessage, err);
      throw new UpdateSessionError(`Error updating session with sessionId: ${sessionId}`);
    }
  });
}
function fetchStatusUrl(sessionId) {
  return __async(this, null, function* () {
    validateFunctionParams(
      [{ input: sessionId, paramName: "sessionId", isString: true }],
      "fetchStatusUrl"
    );
    try {
      const response = yield http.client(`${constants.DEFAULT_RECLAIM_STATUS_URL}${sessionId}`, {
        method: "GET",
        headers: { "Content-Type": "application/json" }
      });
      const res = yield response.json();
      if (!response.ok) {
        const errorMessage = `Error fetching status URL for sessionId: ${sessionId}. Status Code: ${response.status}`;
        logger4.info(errorMessage, res);
        throw new StatusUrlError(errorMessage);
      }
      return res;
    } catch (err) {
      const errorMessage = `Failed to fetch status URL for sessionId: ${sessionId}`;
      logger4.info(errorMessage, err);
      throw new StatusUrlError(`Error fetching status URL for sessionId: ${sessionId}`);
    }
  });
}

// src/utils/proofUtils.ts
var import_ethers3 = require("ethers");

// src/witness.ts
var import_ethers2 = require("ethers");
function createSignDataForClaim(data) {
  const identifier = getIdentifierFromClaimInfo(data);
  const lines = [
    identifier,
    data.owner.toLowerCase(),
    data.timestampS.toString(),
    data.epoch.toString()
  ];
  return lines.join("\n");
}
function getIdentifierFromClaimInfo(info) {
  let canonicalContext = info.context || "";
  if (canonicalContext.length > 0) {
    try {
      const ctx = JSON.parse(canonicalContext);
      canonicalContext = canonicalStringify(ctx);
    } catch (e) {
      throw new Error("unable to parse non-empty context. Must be JSON");
    }
  }
  const str = `${info.provider}
${info.parameters}
${canonicalContext}`;
  return import_ethers2.ethers.keccak256(strToUint8Array(str)).toLowerCase();
}
function strToUint8Array(str) {
  return new TextEncoder().encode(str);
}

// src/utils/proofUtils.ts
var logger5 = logger_default.logger;
function getShortenedUrl(url) {
  return __async(this, null, function* () {
    logger5.info(`Attempting to shorten URL: ${url}`);
    try {
      validateURL(url, "getShortenedUrl");
      const response = yield http.client(`${BACKEND_BASE_URL}/api/sdk/shortener`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ fullUrl: url })
      });
      const res = yield response.json();
      if (!response.ok) {
        logger5.info(`Failed to shorten URL: ${url}, Response: ${JSON.stringify(res)}`);
        return url;
      }
      const shortenedVerificationUrl = res.result.shortUrl;
      return shortenedVerificationUrl;
    } catch (err) {
      logger5.info(`Error shortening URL: ${url}, Error: ${err}`);
      return url;
    }
  });
}
function createLinkWithTemplateData(templateData, sharePagePath) {
  return __async(this, null, function* () {
    let template = encodeURIComponent(JSON.stringify(templateData));
    template = replaceAll(template, "(", "%28");
    template = replaceAll(template, ")", "%29");
    const fullLink = sharePagePath ? `${sharePagePath}/?template=${template}` : `${constants.RECLAIM_SHARE_URL}${template}`;
    try {
      const shortenedLink = yield getShortenedUrl(fullLink);
      return shortenedLink;
    } catch (err) {
      logger5.info(`Error creating link for sessionId: ${templateData.sessionId}, Error: ${err}`);
      return fullLink;
    }
  });
}
function getAttestors() {
  return __async(this, null, function* () {
    var _a;
    const response = yield http.client(constants.DEFAULT_ATTESTORS_URL);
    if (!response.ok) {
      (_a = response.body) == null ? void 0 : _a.cancel();
      throw new BackendServerError(
        `Failed to fetch witness addresses: ${response.status}`
      );
    }
    const { data } = yield response.json();
    return data.map((wt) => ({ id: wt.address, url: "" }));
  });
}
function recoverSignersOfSignedClaim({
  claim,
  signatures
}) {
  const dataStr = createSignDataForClaim(__spreadValues({}, claim));
  const signers = signatures.map(
    (signature) => import_ethers3.ethers.verifyMessage(dataStr, import_ethers3.ethers.hexlify(signature)).toLowerCase()
  );
  return signers;
}

// src/utils/modalUtils.ts
var logger6 = logger_default.logger;
var QRCodeModal = class {
  constructor(options = {}) {
    this.countdownSeconds = 60;
    this.modalId = "reclaim-qr-modal";
    this.options = __spreadValues({
      title: "Verify with Reclaim",
      description: "Scan the QR code with your mobile device to complete verification",
      extensionUrl: constants.CHROME_EXTENSION_URL,
      darkTheme: false,
      modalPopupTimer: 1,
      // default to 1 minute
      showExtensionInstallButton: false
    }, options);
  }
  show(requestUrl) {
    return __async(this, null, function* () {
      try {
        this.close();
        const modalHTML = this.createModalHTML();
        document.body.insertAdjacentHTML("beforeend", modalHTML);
        yield this.generateQRCode(requestUrl, "reclaim-qr-code");
        this.addEventListeners();
        this.startAutoCloseTimer();
      } catch (error) {
        logger6.info("Error showing QR code modal:", error);
        throw error;
      }
    });
  }
  close() {
    if (this.autoCloseTimer) {
      clearTimeout(this.autoCloseTimer);
      this.autoCloseTimer = void 0;
    }
    if (this.countdownTimer) {
      clearInterval(this.countdownTimer);
      this.countdownTimer = void 0;
    }
    const modal = document.getElementById(this.modalId);
    if (modal) {
      modal.remove();
    }
    if (this.options.onClose) {
      this.options.onClose();
    }
  }
  getThemeStyles() {
    const isDark = this.options.darkTheme;
    return {
      modalBackground: isDark ? "rgba(0, 0, 0, 0.8)" : "rgba(0, 0, 0, 0.5)",
      cardBackground: isDark ? "#1f2937" : "white",
      titleColor: isDark ? "#f9fafb" : "#1f2937",
      textColor: isDark ? "#d1d5db" : "#6b7280",
      qrBackground: isDark ? "#374151" : "#f9fafb",
      tipBackground: isDark ? "#1e40af" : "#f0f9ff",
      tipBorder: isDark ? "#1e40af" : "#e0f2fe",
      tipTextColor: isDark ? "#dbeafe" : "#0369a1",
      buttonBackground: isDark ? "#374151" : "#f3f4f6",
      buttonColor: isDark ? "#f9fafb" : "#374151",
      buttonHoverBackground: isDark ? "#4b5563" : "#e5e7eb",
      countdownColor: isDark ? "#6b7280" : "#9ca3af",
      progressBackground: isDark ? "#4b5563" : "#e5e7eb",
      progressGradient: isDark ? "linear-gradient(90deg, #3b82f6 0%, #2563eb 50%, #1d4ed8 100%)" : "linear-gradient(90deg, #2563eb 0%, #1d4ed8 50%, #1e40af 100%)",
      linkColor: isDark ? "#60a5fa" : "#2563eb",
      extensionButtonBackground: isDark ? "#1e40af" : "#2563eb",
      extensionButtonHover: isDark ? "#1d4ed8" : "#1d4ed8"
    };
  }
  createModalHTML() {
    const styles = this.getThemeStyles();
    return `
            <div id="${this.modalId}" style="
                position: fixed;
                top: 0;
                left: 0;
                width: 100%;
                height: 100%;
                background-color: ${styles.modalBackground};
                display: flex;
                justify-content: center;
                align-items: center;
                z-index: 10000;
                font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif;
            ">
                <div style="
                    background: ${styles.cardBackground};
                    border-radius: 12px;
                    padding: 32px;
                    max-width: 400px;
                    width: 90%;
                    text-align: center;
                    box-shadow: 0 20px 25px -5px rgba(0, 0, 0, 0.1), 0 10px 10px -5px rgba(0, 0, 0, 0.04);
                    position: relative;
                ">
                    <button id="reclaim-close-modal" style="
                        position: absolute;
                        top: 16px;
                        right: 16px;
                        background: none;
                        border: none;
                        cursor: pointer;
                        padding: 4px;
                        border-radius: 6px;
                        display: flex;
                        align-items: center;
                        justify-content: center;
                        transition: background-color 0.2s;
                        width: 32px;
                        height: 32px;
                    "
                    onmouseover="this.style.backgroundColor='${styles.buttonHoverBackground}'"
                    onmouseout="this.style.backgroundColor='transparent'"
                    title="Close modal">
                        <svg width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
                            <path d="M12 4L4 12M4 4L12 12" stroke="${styles.buttonColor}" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"/>
                        </svg>
                    </button>
                    
                    <h2 style="
                        margin: 0 0 16px 0;
                        font-size: 24px;
                        font-weight: 600;
                        color: ${styles.titleColor};
                    ">${this.options.title}</h2>
                    
                    <p style="
                        margin: 0 0 24px 0;
                        color: ${styles.textColor};
                        font-size: 14px;
                        line-height: 1.5;
                    ">${this.options.description}</p>
                    
                    <div id="reclaim-qr-code" style="
                        margin: 0 auto 24px auto;
                        background: ${styles.qrBackground};
                        border-radius: 8px;
                        display: inline-block;
                    "></div>
                    
                    ${this.options.showExtensionInstallButton ? `
                    <div style="
                        margin-bottom: 24px;
                        padding: 16px;
                        background: ${styles.tipBackground};
                        border: 1px solid ${styles.tipBorder};
                        border-radius: 8px;
                    ">
                        <p style="
                            margin: 0 0 12px 0;
                            font-size: 14px;
                            color: ${styles.tipTextColor};
                            font-weight: 500;
                        ">\u{1F4A1} For a better experience</p>
                        <p style="
                            margin: 0 0 12px 0;
                            font-size: 13px;
                            color: ${styles.tipTextColor};
                            line-height: 1.4;
                        ">Install our browser extension for seamless verification without QR codes</p>
                        <a href="${this.options.extensionUrl}" 
                           target="_blank" 
                           style="
                               display: inline-block;
                               background: ${styles.extensionButtonBackground};
                               color: white;
                               text-decoration: none;
                               padding: 8px 16px;
                               border-radius: 6px;
                               font-size: 12px;
                               font-weight: 500;
                               transition: background-color 0.2s;
                           "
                           onmouseover="this.style.backgroundColor='${styles.extensionButtonHover}'"
                           onmouseout="this.style.backgroundColor='${styles.extensionButtonBackground}'">
                            Install Extension
                        </a>
                    </div>` : ""}
                    
                    <div style="margin-top: 16px;">
                        <div id="reclaim-countdown" style="
                            font-size: 12px;
                            color: ${styles.countdownColor};
                            font-weight: 400;
                            margin-bottom: 8px;
                        ">Auto-close in 1:00</div>
                        
                        <div style="
                            width: 100%;
                            height: 4px;
                            background-color: ${styles.progressBackground};
                            border-radius: 2px;
                            overflow: hidden;
                        ">
                            <div id="reclaim-progress-bar" style="
                                width: 100%;
                                height: 100%;
                                background: ${styles.progressGradient};
                                border-radius: 2px;
                                transition: width 1s linear;
                            "></div>
                        </div>
                    </div>
                </div>
            </div>
        `;
  }
  generateQRCode(text, containerId) {
    return __async(this, null, function* () {
      try {
        const qrCodeUrl = `${constants.QR_CODE_API_URL}?size=200x200&data=${encodeURIComponent(text)}`;
        const container = document.getElementById(containerId);
        const styles = this.getThemeStyles();
        if (container) {
          container.innerHTML = `
                    <img src="${qrCodeUrl}" 
                         alt="QR Code for Reclaim verification" 
                         style="width: 200px; height: 200px; border-radius: 4px;"
                         onerror="this.style.display='none'; this.nextElementSibling.style.display='block';">
                    <div style="display: none; padding: 20px; color: ${styles.textColor}; font-size: 14px;">
                        QR code could not be loaded.<br>
                        <a href="${text}" target="_blank" style="color: ${styles.linkColor}; text-decoration: underline;">
                            Click here to open verification link
                        </a>
                    </div>
                `;
        }
      } catch (error) {
        logger6.info("Error generating QR code:", error);
        const container = document.getElementById(containerId);
        const styles = this.getThemeStyles();
        if (container) {
          container.innerHTML = `
                    <div style="padding: 20px; color: ${styles.textColor}; font-size: 14px;">
                        <a href="${text}" target="_blank" style="color: ${styles.linkColor}; text-decoration: underline;">
                            Click here to open verification link
                        </a>
                    </div>
                `;
        }
      }
    });
  }
  addEventListeners() {
    const closeButton = document.getElementById("reclaim-close-modal");
    const modal = document.getElementById(this.modalId);
    const closeModal = () => {
      this.close();
    };
    if (closeButton) {
      closeButton.addEventListener("click", closeModal);
    }
    if (modal) {
      modal.addEventListener("click", (e) => {
        if (e.target === modal) {
          closeModal();
        }
      });
    }
    const handleEscape = (e) => {
      if (e.key === "Escape") {
        closeModal();
        document.removeEventListener("keydown", handleEscape);
      }
    };
    document.addEventListener("keydown", handleEscape);
  }
  startAutoCloseTimer() {
    this.countdownSeconds = (this.options.modalPopupTimer || 1) * 60;
    this.updateCountdownDisplay();
    this.countdownTimer = setInterval(() => {
      this.countdownSeconds--;
      this.updateCountdownDisplay();
      if (this.countdownSeconds <= 0) {
        this.close();
      }
    }, 1e3);
    const autoCloseMs = (this.options.modalPopupTimer || 1) * 60 * 1e3;
    this.autoCloseTimer = setTimeout(() => {
      this.close();
    }, autoCloseMs);
  }
  updateCountdownDisplay() {
    const countdownElement = document.getElementById("reclaim-countdown");
    const progressBar = document.getElementById("reclaim-progress-bar");
    if (countdownElement) {
      const minutes = Math.floor(this.countdownSeconds / 60);
      const seconds = this.countdownSeconds % 60;
      const timeString = `${minutes}:${seconds.toString().padStart(2, "0")}`;
      countdownElement.textContent = `Auto-close in ${timeString}`;
    }
    if (progressBar) {
      const totalSeconds = (this.options.modalPopupTimer || 1) * 60;
      const progressPercentage = this.countdownSeconds / totalSeconds * 100;
      progressBar.style.width = `${progressPercentage}%`;
    }
  }
};

// src/utils/device.ts
var navigatorDefined = typeof navigator !== "undefined";
var windowDefined = typeof window !== "undefined";
var userAgent = navigatorDefined ? navigator.userAgent.toLowerCase() : "";
var userAgentData = navigatorDefined ? navigator.userAgentData : void 0;
var cachedDeviceType = null;
var cachedMobileType = null;
function safeMatchMedia(query) {
  var _a, _b;
  try {
    return ((_b = (_a = window.matchMedia) == null ? void 0 : _a.call(window, query)) == null ? void 0 : _b.matches) || false;
  } catch (e) {
    return false;
  }
}
function safeCSSSupports(property, value) {
  var _a;
  try {
    return ((_a = CSS == null ? void 0 : CSS.supports) == null ? void 0 : _a.call(CSS, property, value)) || false;
  } catch (e) {
    return false;
  }
}
function safeQuerySelector(selector) {
  var _a;
  try {
    return ((_a = document == null ? void 0 : document.querySelector) == null ? void 0 : _a.call(document, selector)) !== null;
  } catch (e) {
    return false;
  }
}
function getDeviceType() {
  var _a, _b;
  if (cachedDeviceType !== null) {
    return cachedDeviceType;
  }
  if (!navigatorDefined || !windowDefined) {
    return "desktop" /* DESKTOP */;
  }
  let mobileScore = 0;
  const CONFIDENCE_THRESHOLD = 3;
  const screenWidth = window.innerWidth || ((_a = window.screen) == null ? void 0 : _a.width) || 0;
  const screenHeight = window.innerHeight || ((_b = window.screen) == null ? void 0 : _b.height) || 0;
  const hasSmallScreen = screenWidth <= 480 || screenHeight <= 480;
  const hasLargeScreen = screenWidth > 1024 && screenHeight > 768;
  const hasTouch = "ontouchstart" in window || navigatorDefined && navigator.maxTouchPoints > 0;
  const hasPreciseMouse = safeMatchMedia("(pointer: fine)");
  const canHover = safeMatchMedia("(hover: hover)");
  const hasMouseAndTouch = hasTouch && hasPreciseMouse;
  const isWindowsTouchLaptop = /Windows/i.test(userAgent) && hasPreciseMouse && hasTouch;
  if (hasTouch && !hasMouseAndTouch) {
    mobileScore += 2;
  } else if (hasMouseAndTouch) {
    mobileScore += 1;
  }
  if (hasSmallScreen && !isWindowsTouchLaptop) {
    mobileScore += 2;
  }
  const hasMobileUserAgent = /android|webos|iphone|ipad|ipod|blackberry|iemobile|opera mini|mobile/i.test(userAgent);
  if (hasMobileUserAgent) {
    mobileScore += 3;
  }
  const hasMobileAPIs = "orientation" in window || "DeviceMotionEvent" in window || "DeviceOrientationEvent" in window;
  const isDesktopSafari = /Safari/i.test(userAgent) && !/Mobile/i.test(userAgent) && /Mac|Intel/i.test(userAgent);
  if (hasMobileAPIs && (hasSmallScreen || hasMobileUserAgent) && !isDesktopSafari && !isWindowsTouchLaptop) {
    mobileScore += 2;
  }
  const hasHighDPI = window.devicePixelRatio > 1.5;
  if (hasHighDPI && hasSmallScreen) {
    mobileScore += 1;
  }
  const hasViewportMeta = safeQuerySelector('meta[name="viewport"]');
  if (hasViewportMeta && hasSmallScreen) {
    mobileScore += 1;
  }
  const isPadProInDesktopMode = userAgent.includes("macintosh") && hasTouch;
  if (isPadProInDesktopMode) {
    mobileScore += 2;
  }
  if (hasLargeScreen && hasPreciseMouse) {
    mobileScore -= 3;
  } else if (hasLargeScreen && !hasTouch) {
    mobileScore -= 2;
  }
  if (hasPreciseMouse && canHover) {
    mobileScore -= 2;
  }
  const isWindowsDesktop = /Windows/i.test(userAgent) && !hasMobileUserAgent;
  if (isWindowsDesktop) {
    mobileScore -= 3;
  }
  cachedDeviceType = mobileScore >= CONFIDENCE_THRESHOLD ? "mobile" /* MOBILE */ : "desktop" /* DESKTOP */;
  return cachedDeviceType;
}
function getMobileDeviceType() {
  var _a;
  if (cachedMobileType !== null) {
    return cachedMobileType;
  }
  if (!navigatorDefined || !windowDefined) {
    return "android" /* ANDROID */;
  }
  const ua = navigator.userAgent;
  const hasIOSDeviceName = /iPad|iPhone|iPod/i.test(ua);
  if (hasIOSDeviceName) {
    cachedMobileType = "ios" /* IOS */;
    return cachedMobileType;
  }
  const isMacWithTouch = /Macintosh|MacIntel/i.test(ua) && "ontouchstart" in window;
  const isMacOSWithTouch = (userAgentData == null ? void 0 : userAgentData.platform) === "macOS" && "ontouchstart" in window;
  if (isMacWithTouch || isMacOSWithTouch) {
    cachedMobileType = "ios" /* IOS */;
    return cachedMobileType;
  }
  const hasIOSPermissionAPI = typeof ((_a = window.DeviceMotionEvent) == null ? void 0 : _a.requestPermission) === "function";
  const hasIOSTouchCallout = safeCSSSupports("-webkit-touch-callout", "none");
  if (hasIOSPermissionAPI || hasIOSTouchCallout) {
    cachedMobileType = "ios" /* IOS */;
    return cachedMobileType;
  }
  const isIOSWebKit = /WebKit/i.test(ua) && !/Chrome|CriOS|Android/i.test(ua) && !/Macintosh|MacIntel/i.test(ua);
  if (isIOSWebKit) {
    cachedMobileType = "ios" /* IOS */;
    return cachedMobileType;
  }
  const hasAndroidKeyword = /Android/i.test(ua);
  if (hasAndroidKeyword) {
    cachedMobileType = "android" /* ANDROID */;
    return cachedMobileType;
  }
  const isChromeOnMobile = window.chrome && /Mobile/i.test(ua);
  if (isChromeOnMobile) {
    cachedMobileType = "android" /* ANDROID */;
    return cachedMobileType;
  }
  cachedMobileType = "android" /* ANDROID */;
  return cachedMobileType;
}
function isMobileDevice() {
  return getDeviceType() === "mobile" /* MOBILE */;
}
function isDesktopDevice() {
  return getDeviceType() === "desktop" /* DESKTOP */;
}
function clearDeviceCache() {
  cachedDeviceType = null;
  cachedMobileType = null;
}

// src/Reclaim.ts
var logger7 = logger_default.logger;
var sdkVersion = require_package().version;
function verifyProof(proofOrProofs, allowAiWitness = false) {
  return __async(this, null, function* () {
    try {
      yield assertValidProof(proofOrProofs, allowAiWitness);
      return true;
    } catch (error) {
      logger7.error("error in validating proof", error);
      return false;
    }
  });
}
function assertValidProof(proofOrProofs, allowAiWitness = false) {
  return __async(this, null, function* () {
    const attestors = yield getAttestors();
    proofOrProofs = Array.isArray(proofOrProofs) ? proofOrProofs : [proofOrProofs];
    for (const proof of proofOrProofs) {
      const signers = recoverSignersOfSignedClaim({
        claim: proof.claimData,
        signatures: proof.signatures.map((signature) => import_ethers4.ethers.getBytes(signature))
      });
      if (!attestors.some((attestor) => signers.includes(attestor.id.toLowerCase()))) {
        throw new ProofNotVerifiedError("Identifier mismatch");
      }
    }
  });
}
function transformForOnchain(proof) {
  const claimInfoBuilder = /* @__PURE__ */ new Map([
    ["context", proof.claimData.context],
    ["parameters", proof.claimData.parameters],
    ["provider", proof.claimData.provider]
  ]);
  const claimInfo = Object.fromEntries(claimInfoBuilder);
  const claimBuilder = /* @__PURE__ */ new Map([
    ["epoch", proof.claimData.epoch],
    ["identifier", proof.claimData.identifier],
    ["owner", proof.claimData.owner],
    ["timestampS", proof.claimData.timestampS]
  ]);
  const signedClaim = {
    claim: Object.fromEntries(claimBuilder),
    signatures: proof.signatures
  };
  return { claimInfo, signedClaim };
}
var emptyTemplateData = {
  sessionId: "",
  providerId: "",
  applicationId: "",
  signature: "",
  timestamp: "",
  callbackUrl: "",
  context: "",
  parameters: {},
  redirectUrl: "",
  redirectUrlOptions: { method: "GET" },
  cancelCallbackUrl: "",
  cancelRedirectUrl: "",
  cancelRedirectUrlOptions: { method: "GET" },
  acceptAiProviders: false,
  sdkVersion: "",
  providerVersion: "",
  resolvedProviderVersion: "",
  jsonProofResponse: false,
  log: false
};
var ReclaimProofRequest = class _ReclaimProofRequest {
  // 30 seconds timeout, can be adjusted
  constructor(applicationId, providerId, options) {
    this.context = { contextAddress: "0x0", contextMessage: "sample context" };
    this.claimCreationType = "createClaim" /* STANDALONE */;
    this.intervals = /* @__PURE__ */ new Map();
    this.jsonProofResponse = false;
    this.extensionID = "reclaim-extension";
    this.FAILURE_TIMEOUT = 30 * 1e3;
    /**
     * Validates signature and returns template data
     * @returns
     */
    this.getTemplateData = () => {
      var _a, _b, _c, _d, _e, _f, _g, _h, _i, _j, _k, _l;
      if (!this.signature) {
        throw new SignatureNotFoundError("Signature is not set.");
      }
      const defaultHosts = ["share.reclaimprotocol.org", "portal.reclaimprotocol.org"];
      if (this.customSharePageUrl) {
        try {
          const sharePageHost = new URL(this.customSharePageUrl).hostname;
          if (!defaultHosts.includes(sharePageHost) && !this.appCallbackUrl) {
            throw new CallbackUrlRequiredError(
              "A custom callback URL is required when using a customSharePage url"
            );
          }
        } catch (e) {
          if (e instanceof CallbackUrlRequiredError) throw e;
        }
      }
      validateSignature(this.providerId, this.signature, this.applicationId, this.timeStamp);
      const templateData = {
        sessionId: this.sessionId,
        providerId: this.providerId,
        applicationId: this.applicationId,
        signature: this.signature,
        timestamp: this.timeStamp,
        callbackUrl: this.getAppCallbackUrl(),
        context: canonicalStringify(this.context),
        providerVersion: (_b = (_a = this.options) == null ? void 0 : _a.providerVersion) != null ? _b : "",
        resolvedProviderVersion: (_c = this.resolvedProviderVersion) != null ? _c : "",
        parameters: this.parameters,
        redirectUrl: (_d = this.redirectUrl) != null ? _d : "",
        redirectUrlOptions: this.redirectUrlOptions,
        cancelCallbackUrl: this.getCancelCallbackUrl(),
        cancelRedirectUrl: this.cancelRedirectUrl,
        cancelRedirectUrlOptions: this.cancelRedirectUrlOptions,
        acceptAiProviders: (_f = (_e = this.options) == null ? void 0 : _e.acceptAiProviders) != null ? _f : false,
        sdkVersion: this.sdkVersion,
        jsonProofResponse: this.jsonProofResponse,
        log: (_h = (_g = this.options) == null ? void 0 : _g.log) != null ? _h : false,
        canAutoSubmit: (_j = (_i = this.options) == null ? void 0 : _i.canAutoSubmit) != null ? _j : true,
        metadata: (_k = this.options) == null ? void 0 : _k.metadata,
        preferredLocale: (_l = this.options) == null ? void 0 : _l.preferredLocale
      };
      return templateData;
    };
    var _a, _b;
    this.providerId = providerId;
    this.timeStamp = Date.now().toString();
    this.applicationId = applicationId;
    this.sessionId = "";
    this.templateData = emptyTemplateData;
    this.parameters = {};
    if (!options) {
      options = {};
    }
    options.useBrowserExtension = (_a = options.useBrowserExtension) != null ? _a : true;
    if (options == null ? void 0 : options.log) {
      logger_default.setLogLevel("info");
    } else {
      logger_default.setLogLevel("silent");
    }
    if (options.useAppClip === void 0) {
      options.useAppClip = true;
    }
    if (options == null ? void 0 : options.envUrl) {
      setBackendBaseUrl(options.envUrl);
    } else if ((_b = options == null ? void 0 : options.customSharePageUrl) == null ? void 0 : _b.includes("eu.portal.reclaimprotocol.org")) {
      setBackendBaseUrl("https://eu.api.reclaimprotocol.org");
    }
    if (options.extensionID) {
      this.extensionID = options.extensionID;
    }
    if (options == null ? void 0 : options.customSharePageUrl) {
      this.customSharePageUrl = options.customSharePageUrl;
    }
    if (options == null ? void 0 : options.customAppClipUrl) {
      this.customAppClipUrl = options.customAppClipUrl;
    }
    this.options = options;
    this.sdkVersion = "js-" + sdkVersion;
    logger7.info(`Initializing client with applicationId: ${this.applicationId}`);
  }
  /**
   * Initializes a new Reclaim proof request instance with automatic signature generation and session creation
   *
   * @param applicationId - Your Reclaim application ID
   * @param appSecret - Your application secret key for signing requests
   * @param providerId - The ID of the provider to use for proof generation
   * @param options - Optional configuration options for the proof request
   * @returns Promise<ReclaimProofRequest> - A fully initialized proof request instance
   * @throws {InitError} When initialization fails due to invalid parameters or session creation errors
   *
   * @example
   * ```typescript
   * const proofRequest = await ReclaimProofRequest.init(
   *   'your-app-id',
   *   'your-app-secret',
   *   'provider-id',
   *   { log: true, acceptAiProviders: true }
   * );
   * ```
   */
  static init(applicationId, appSecret, providerId, options) {
    return __async(this, null, function* () {
      try {
        validateFunctionParams([
          { paramName: "applicationId", input: applicationId, isString: true },
          { paramName: "providerId", input: providerId, isString: true },
          { paramName: "appSecret", input: appSecret, isString: true }
        ], "the constructor");
        if (options) {
          if (options.acceptAiProviders) {
            validateFunctionParams([
              { paramName: "acceptAiProviders", input: options.acceptAiProviders }
            ], "the constructor");
          }
          if (options.providerVersion) {
            validateFunctionParams([
              { paramName: "providerVersion", input: options.providerVersion, isString: true }
            ], "the constructor");
          }
          if (options.log) {
            validateFunctionParams([
              { paramName: "log", input: options.log }
            ], "the constructor");
          }
          if (options.useAppClip) {
            validateFunctionParams([
              { paramName: "useAppClip", input: options.useAppClip }
            ], "the constructor");
          }
          if (options.device) {
            validateFunctionParams([
              { paramName: "device", input: options.device, isString: true }
            ], "the constructor");
          }
          if (options.useBrowserExtension) {
            validateFunctionParams([
              { paramName: "useBrowserExtension", input: options.useBrowserExtension }
            ], "the constructor");
          }
          if (options.extensionID) {
            validateFunctionParams([
              { paramName: "extensionID", input: options.extensionID, isString: true }
            ], "the constructor");
          }
          if (options.envUrl) {
            validateFunctionParams([
              { paramName: "envUrl", input: options.envUrl, isString: true }
            ], "the constructor");
          }
          if (options.customSharePageUrl) {
            validateFunctionParams([
              { paramName: "customSharePageUrl", input: options.customSharePageUrl, isString: true }
            ], "the constructor");
          }
          if (options.customAppClipUrl) {
            validateFunctionParams([
              { paramName: "customAppClipUrl", input: options.customAppClipUrl, isString: true }
            ], "the constructor");
          }
          if (options.preferredLocale) {
            validateFunctionParams([
              { paramName: "preferredLocale", input: options.preferredLocale, isString: true }
            ], "the constructor");
            validateFunctionParamsWithFn({
              paramName: "preferredLocale",
              input: options.preferredLocale,
              isValid: () => {
                try {
                  Intl.getCanonicalLocales(options.preferredLocale);
                  return true;
                } catch (error) {
                  logger7.info("Failed to canonicalize locale", error);
                  return false;
                }
              }
            }, "the constructor");
          }
        }
        const proofRequestInstance = new _ReclaimProofRequest(applicationId, providerId, options);
        const signature = yield proofRequestInstance.generateSignature(appSecret);
        proofRequestInstance.setSignature(signature);
        const data = yield initSession(providerId, applicationId, proofRequestInstance.timeStamp, signature, options == null ? void 0 : options.providerVersion);
        proofRequestInstance.sessionId = data.sessionId;
        proofRequestInstance.resolvedProviderVersion = data.resolvedProviderVersion;
        return proofRequestInstance;
      } catch (error) {
        console.error(error);
        logger7.info("Failed to initialize ReclaimProofRequest", error);
        throw new InitError("Failed to initialize ReclaimProofRequest", error);
      }
    });
  }
  /**
   * Creates a ReclaimProofRequest instance from a JSON string representation
   *
   * This method deserializes a previously exported proof request (via toJsonString) and reconstructs
   * the instance with all its properties. Useful for recreating requests on the frontend or across different contexts.
   *
   * @param jsonString - JSON string containing the serialized proof request data
   * @returns {Promise<ReclaimProofRequest>} - Reconstructed proof request instance
   * @throws {InvalidParamError} When JSON string is invalid or contains invalid parameters
   *
   * @example
   * ```typescript
   * const jsonString = proofRequest.toJsonString();
   * const reconstructed = await ReclaimProofRequest.fromJsonString(jsonString);
   * // Can also be used with InApp SDK's startVerificationFromJson method
   * ```
   */
  static fromJsonString(jsonString) {
    return __async(this, null, function* () {
      try {
        const {
          applicationId,
          providerId,
          sessionId,
          context,
          parameters,
          signature,
          redirectUrl,
          redirectUrlOptions,
          cancelCallbackUrl,
          cancelRedirectUrl,
          cancelRedirectUrlOptions,
          timeStamp,
          timestamp,
          appCallbackUrl,
          claimCreationType,
          options,
          sdkVersion: sdkVersion2,
          jsonProofResponse,
          resolvedProviderVersion,
          modalOptions
        } = JSON.parse(jsonString);
        const resolvedTimestamp = timestamp || timeStamp;
        validateFunctionParams([
          { input: applicationId, paramName: "applicationId", isString: true },
          { input: providerId, paramName: "providerId", isString: true },
          { input: signature, paramName: "signature", isString: true },
          { input: sessionId, paramName: "sessionId", isString: true },
          { input: resolvedTimestamp, paramName: "timestamp", isString: true },
          { input: sdkVersion2, paramName: "sdkVersion", isString: true }
        ], "fromJsonString");
        if (modalOptions) {
          validateModalOptions(modalOptions, "fromJsonString", "modalOptions.");
        }
        if (redirectUrl) {
          validateURL(redirectUrl, "fromJsonString");
        }
        if (redirectUrlOptions) {
          validateRedirectionMethod(redirectUrlOptions.method, "fromJsonString");
          validateRedirectionBody(redirectUrlOptions.body, "fromJsonString");
        }
        if (appCallbackUrl) {
          validateURL(appCallbackUrl, "fromJsonString");
        }
        if (cancelRedirectUrl) {
          validateURL(cancelRedirectUrl, "fromJsonString");
        }
        if (cancelRedirectUrlOptions) {
          validateRedirectionMethod(cancelRedirectUrlOptions.method, "fromJsonString");
          validateRedirectionBody(cancelRedirectUrlOptions.body, "fromJsonString");
        }
        if (cancelCallbackUrl) {
          validateURL(cancelCallbackUrl, "fromJsonString");
        }
        if (context) {
          validateContext(context);
        }
        if (parameters) {
          validateParameters(parameters);
        }
        if (claimCreationType) {
          validateFunctionParams([
            { input: claimCreationType, paramName: "claimCreationType" }
          ], "fromJsonString");
        }
        if (jsonProofResponse !== void 0) {
          validateFunctionParams([
            { input: jsonProofResponse, paramName: "jsonProofResponse" }
          ], "fromJsonString");
        }
        if (options == null ? void 0 : options.providerVersion) {
          validateFunctionParams([
            { input: options.providerVersion, paramName: "options.providerVersion", isString: true }
          ], "fromJsonString");
        }
        if (resolvedProviderVersion) {
          validateFunctionParams([
            { input: resolvedProviderVersion, paramName: "resolvedProviderVersion", isString: true }
          ], "fromJsonString");
        }
        if (options == null ? void 0 : options.preferredLocale) {
          validateFunctionParams([
            { paramName: "options.preferredLocale", input: options.preferredLocale, isString: true }
          ], "fromJsonString");
          validateFunctionParamsWithFn({
            paramName: "options.preferredLocale",
            input: options.preferredLocale,
            isValid: () => {
              try {
                Intl.getCanonicalLocales(options.preferredLocale);
                return true;
              } catch (error) {
                logger7.info("Failed to canonicalize locale", error);
                return false;
              }
            }
          }, "fromJsonString");
        }
        const proofRequestInstance = new _ReclaimProofRequest(applicationId, providerId, options);
        proofRequestInstance.sessionId = sessionId;
        proofRequestInstance.context = context;
        proofRequestInstance.parameters = parameters;
        proofRequestInstance.appCallbackUrl = appCallbackUrl;
        proofRequestInstance.redirectUrl = redirectUrl;
        proofRequestInstance.redirectUrlOptions = redirectUrlOptions;
        proofRequestInstance.timeStamp = resolvedTimestamp;
        proofRequestInstance.signature = signature;
        proofRequestInstance.sdkVersion = sdkVersion2;
        proofRequestInstance.resolvedProviderVersion = resolvedProviderVersion;
        proofRequestInstance.modalOptions = modalOptions;
        proofRequestInstance.jsonProofResponse = jsonProofResponse != null ? jsonProofResponse : false;
        proofRequestInstance.cancelCallbackUrl = cancelCallbackUrl;
        proofRequestInstance.cancelRedirectUrl = cancelRedirectUrl;
        proofRequestInstance.cancelRedirectUrlOptions = cancelRedirectUrlOptions;
        return proofRequestInstance;
      } catch (error) {
        logger7.info("Failed to parse JSON string in fromJsonString:", error);
        throw new InvalidParamError("Invalid JSON string provided to fromJsonString");
      }
    });
  }
  /**
   * Sets a custom callback URL where proofs will be submitted via HTTP `POST`
   *
   * By default, proofs are sent as HTTP POST with `Content-Type` as `application/x-www-form-urlencoded`.
   * Pass function argument `jsonProofResponse` as `true` to send proofs with `Content-Type` as `application/json`.
   *
   * When a custom callback URL is set, proofs are sent to the custom URL *instead* of the Reclaim backend.
   * Consequently, the startSession `onSuccess` callback will be invoked with an empty array (`[]`)
   * instead of the proof data, as the proof is not available to the SDK in this flow.
   *
   * This verification session's id will be present in `X-Reclaim-Session-Id` header of the request.
   * The request URL will contain query param `allowAiWitness` with value `true` when AI Witness should be allowed by handler of the request.
   *
   * Note: InApp SDKs are unaffected by this property as they do not handle proof submission.
   *
   * @param url - The URL where proofs should be submitted via HTTP `POST`
   * @param jsonProofResponse - Optional. Set to true to submit proofs as `application/json`. Defaults to false
   * @throws {InvalidParamError} When URL is invalid
   *
   * @example
   * ```typescript
   * proofRequest.setAppCallbackUrl('https://your-backend.com/callback');
   * // Or with JSON format
   * proofRequest.setAppCallbackUrl('https://your-backend.com/callback', true);
   * ```
   */
  setAppCallbackUrl(url, jsonProofResponse) {
    validateURL(url, "setAppCallbackUrl");
    this.appCallbackUrl = url;
    this.jsonProofResponse = jsonProofResponse != null ? jsonProofResponse : false;
  }
  /**
   * Sets a redirect URL where users will be redirected after successfully acquiring and submitting proof
   *
   * @param url - The URL where users should be redirected after successful proof generation
   * @param method - The redirection method that should be used for redirection. Allowed options: `GET`, and `POST`.
   * `POST` form redirection is only supported in In-Browser SDK.
   * @param body - List of name-value pairs to be sent as the body of the form request.
   * `When `method` is set to `POST`, `body` will be sent with 'application/x-www-form-urlencoded' content type.
   * When `method` is set to `GET`, if `body` is set then `body` will be sent as query parameters.
   * Sending `body` on redirection is only supported in In-Browser SDK.
   *
   * @throws {InvalidParamError} When URL is invalid
   *
   * @example
   * ```typescript
   * proofRequest.setRedirectUrl('https://your-app.com/success');
   * ```
   */
  setRedirectUrl(url, method = "GET", body) {
    validateURL(url, "setRedirectUrl");
    validateRedirectionMethod(method, "setRedirectUrl");
    validateRedirectionBody(body, "setRedirectUrl");
    this.redirectUrl = url;
    this.redirectUrlOptions = { method: method || "GET", body };
  }
  /**
       * Sets a custom callback URL where errors that abort the verification process will be submitted via HTTP POST
       *
       * Errors will be HTTP POSTed with `header 'Content-Type': 'application/json'`.
       * When a custom error callback URL is set, Reclaim will no longer receive errors upon submission,
       * and listeners on the startSession method will not be triggered. Your application must
       * coordinate with your backend to receive errors.
       *
       * This verification session's id will be present in `X-Reclaim-Session-Id` header of the request.
       *
       * Following is the data format which is sent as an HTTP POST request to the url with `Content-Type: application/json`:
  
       * ```json
       * {
       *  "type": "string", // Name of the exception
       *  "message": "string",
       *  "sessionId": "string",
       *   // context as canonicalized json string
       *   "context": "string",
       *   // Other fields with more details about error may be present
       *   // [key: any]: any
       * }
       * ```
       *
       * For more details about response format, check out [official documentation of Error Callback URL](https://docs.reclaimprotocol.org/js-sdk/preparing-request#cancel-callback).
       *
       * @param url - The URL where errors should be submitted via HTTP POST
       * @throws {InvalidParamError} When URL is invalid
       *
       * @example
       * ```typescript
       * proofRequest.setCancelCallbackUrl('https://your-backend.com/error-callback');
       * ```
       *
       * @since 4.8.1
       *
       */
  setCancelCallbackUrl(url) {
    validateURL(url, "setCancelCallbackUrl");
    this.cancelCallbackUrl = url;
  }
  /**
   * Sets an error redirect URL where users will be redirected after an error which aborts the verification process
   *
   * @param url - The URL where users should be redirected after an error which aborts the verification process
   * @param method - The redirection method that should be used for redirection. Allowed options: `GET`, and `POST`.
   * `POST` form redirection is only supported in In-Browser SDK.
   * @param body - List of name-value pairs to be sent as the body of the form request.
   * When `method` is set to `POST`, `body` will be sent with 'application/x-www-form-urlencoded' content type.
   * When `method` is set to `GET`, if `body` is set then `body` will be sent as query parameters.
   * Sending `body` on redirection is only supported in In-Browser SDK.
   * @throws {InvalidParamError} When URL is invalid
   *
   * @example
   * ```typescript
   * proofRequest.setCancelRedirectUrl('https://your-app.com/error');
   * ```
   *
   * @since 4.10.0
   *
   */
  setCancelRedirectUrl(url, method = "GET", body) {
    validateURL(url, "setCancelRedirectUrl");
    validateRedirectionMethod(method, "setCancelRedirectUrl");
    validateRedirectionBody(body, "setCancelRedirectUrl");
    this.cancelRedirectUrl = url;
    this.cancelRedirectUrlOptions = { method: method || "GET", body };
  }
  /**
   * Sets the claim creation type for the proof request
   *
   * @param claimCreationType - The type of claim creation (e.g., STANDALONE)
   *
   * @example
   * ```typescript
   * proofRequest.setClaimCreationType(ClaimCreationType.STANDALONE);
   * ```
   */
  setClaimCreationType(claimCreationType) {
    this.claimCreationType = claimCreationType;
  }
  /**
   * Sets custom options for the QR code modal display
   *
   * @param options - Modal configuration options including title, description, theme, etc.
   * @throws {SetParamsError} When modal options are invalid
   *
   * @example
   * ```typescript
   * proofRequest.setModalOptions({
   *   title: 'Scan QR Code',
   *   description: 'Scan with your mobile device',
   *   darkTheme: true
   * });
   * ```
   */
  setModalOptions(options) {
    try {
      validateModalOptions(options, "setModalOptions");
      this.modalOptions = __spreadValues(__spreadValues({}, this.modalOptions), options);
      logger7.info("Modal options set successfully");
    } catch (error) {
      logger7.info("Error setting modal options:", error);
      throw new SetParamsError("Error setting modal options", error);
    }
  }
  /**
   * Sets additional context data to be stored with the claim
   *
   * This allows you to associate custom JSON serializable data with the proof claim.
   * The context can be retrieved and validated when verifying the proof.
   *
   * Also see [setContext] which is an alternate way to set context that has an address & message.
   *
   * [setContext] and [setJsonContext] overwrite each other. Each call replaces the existing context.
   *
   * @param context - Any additional data you want to store with the claim. Should be serializable to a JSON string.
   * @throws {SetContextError} When context parameters are invalid
   *
   * @example
   * ```typescript
   * proofRequest.setJsonContext({foo: 'bar'});
   * ```
   */
  setJsonContext(context) {
    try {
      validateFunctionParams([
        { input: context, paramName: "context", isString: false }
      ], "setJsonContext");
      this.context = JSON.parse(canonicalStringify(context));
    } catch (error) {
      logger7.info("Error setting context", error);
      throw new SetContextError("Error setting context", error);
    }
  }
  /**
   * Sets additional context data to be stored with the claim
   *
   * This allows you to associate custom data (address and message) with the proof claim.
   * The context can be retrieved and validated when verifying the proof.
   *
   * Also see [setJsonContext] which is an alternate way to set context that allows for custom JSON serializable data.
   *
   * [setContext] and [setJsonContext] overwrite each other. Each call replaces the existing context.
   *
   * @param address - Context address identifier
   * @param message - Additional data to associate with the address
   * @throws {SetContextError} When context parameters are invalid
   *
   * @example
   * ```typescript
   * proofRequest.setContext('0x1234...', 'User verification for premium access');
   * ```
   */
  setContext(address, message) {
    try {
      validateFunctionParams([
        { input: address, paramName: "address", isString: true },
        { input: message, paramName: "message", isString: true }
      ], "setContext");
      this.context = { contextAddress: address, contextMessage: message };
    } catch (error) {
      logger7.info("Error setting context", error);
      throw new SetContextError("Error setting context", error);
    }
  }
  /**
   * @deprecated use setContext instead
   *
   * @param address
   * @param message additional data you want associated with the [address]
   */
  addContext(address, message) {
    this.setContext(address, message);
  }
  /**
   * Sets provider-specific parameters for the proof request
   *
   * These parameters are passed to the provider and may include configuration options,
   * filters, or other provider-specific settings required for proof generation.
   *
   * @param params - Key-value pairs of parameters to set
   * @throws {SetParamsError} When parameters are invalid
   *
   * @example
   * ```typescript
   * proofRequest.setParams({
   *   minFollowers: '1000',
   *   platform: 'twitter'
   * });
   * ```
   */
  setParams(params) {
    try {
      validateParameters(params);
      this.parameters = __spreadValues(__spreadValues({}, this.parameters), params);
    } catch (error) {
      logger7.info("Error Setting Params:", error);
      throw new SetParamsError("Error setting params", error);
    }
  }
  /**
   * Returns the currently configured app callback URL
   *
   * If no custom callback URL was set via setAppCallbackUrl(), this returns the default
   * Reclaim service callback URL with the current session ID.
   *
   * @returns The callback URL where proofs will be submitted
   * @throws {GetAppCallbackUrlError} When unable to retrieve the callback URL
   *
   * @example
   * ```typescript
   * const callbackUrl = proofRequest.getAppCallbackUrl();
   * console.log('Proofs will be sent to:', callbackUrl);
   * ```
   */
  getAppCallbackUrl() {
    try {
      validateFunctionParams([{ input: this.sessionId, paramName: "sessionId", isString: true }], "getAppCallbackUrl");
      return this.appCallbackUrl || `${constants.DEFAULT_RECLAIM_CALLBACK_URL}${this.sessionId}`;
    } catch (error) {
      logger7.info("Error getting app callback url", error);
      throw new GetAppCallbackUrlError("Error getting app callback url", error);
    }
  }
  /**
   * Returns the currently configured cancel callback URL
   *
   * If no custom cancel callback URL was set via setCancelCallbackUrl(), this returns the default
   * Reclaim service cancel callback URL with the current session ID.
   *
   * @returns The cancel callback URL where proofs will be submitted
   * @throws {GetAppCallbackUrlError} When unable to retrieve the cancel callback URL
   *
   * @example
   * ```typescript
   * const callbackUrl = proofRequest.getCancelCallbackUrl();
   * console.log('Errors will be sent to:', callbackUrl);
   * ```
   */
  getCancelCallbackUrl() {
    try {
      validateFunctionParams([{ input: this.sessionId, paramName: "sessionId", isString: true }], "getCancelCallbackUrl");
      return this.cancelCallbackUrl || `${constants.DEFAULT_RECLAIM_CANCEL_CALLBACK_URL}${this.sessionId}`;
    } catch (error) {
      logger7.info("Error getting cancel callback url", error);
      throw new GetAppCallbackUrlError("Error getting cancel callback url", error);
    }
  }
  /**
   * Returns the status URL for monitoring the current session
   *
   * This URL can be used to check the status of the proof request session.
   *
   * @returns The status monitoring URL for the current session
   * @throws {GetStatusUrlError} When unable to retrieve the status URL
   *
   * @example
   * ```typescript
   * const statusUrl = proofRequest.getStatusUrl();
   * // Use this URL to poll for session status updates
   * ```
   */
  getStatusUrl() {
    try {
      validateFunctionParams([{ input: this.sessionId, paramName: "sessionId", isString: true }], "getStatusUrl");
      return `${constants.DEFAULT_RECLAIM_STATUS_URL}${this.sessionId}`;
    } catch (error) {
      logger7.info("Error fetching Status Url", error);
      throw new GetStatusUrlError("Error fetching status url", error);
    }
  }
  /**
   * Returns the session ID associated with this proof request
   *
   * The session ID is automatically generated during initialization and uniquely
   * identifies this proof request session.
   *
   * @returns The session ID string
   * @throws {SessionNotStartedError} When session ID is not set
   *
   * @example
   * ```typescript
   * const sessionId = proofRequest.getSessionId();
   * console.log('Session ID:', sessionId);
   * ```
   */
  getSessionId() {
    if (!this.sessionId) {
      throw new SessionNotStartedError("SessionId is not set");
    }
    return this.sessionId;
  }
  // Private helper methods
  setSignature(signature) {
    try {
      validateFunctionParams([{ input: signature, paramName: "signature", isString: true }], "setSignature");
      this.signature = signature;
      logger7.info(`Signature set successfully for applicationId: ${this.applicationId}`);
    } catch (error) {
      logger7.info("Error setting signature", error);
      throw new SetSignatureError("Error setting signature", error);
    }
  }
  generateSignature(applicationSecret) {
    return __async(this, null, function* () {
      try {
        const wallet = new import_ethers4.ethers.Wallet(applicationSecret);
        const canonicalData = (0, import_canonicalize3.default)({ providerId: this.providerId, timestamp: this.timeStamp });
        if (!canonicalData) {
          throw new SignatureGeneratingError("Failed to canonicalize data for signing.");
        }
        const messageHash = import_ethers4.ethers.keccak256(new TextEncoder().encode(canonicalData));
        return yield wallet.signMessage(import_ethers4.ethers.getBytes(messageHash));
      } catch (err) {
        logger7.info(`Error generating proof request for applicationId: ${this.applicationId}, providerId: ${this.providerId}, signature: ${this.signature}, timeStamp: ${this.timeStamp}`, err);
        throw new SignatureGeneratingError(`Error generating signature for applicationSecret: ${applicationSecret}`);
      }
    });
  }
  clearInterval() {
    if (this.sessionId && this.intervals.has(this.sessionId)) {
      clearInterval(this.intervals.get(this.sessionId));
      this.intervals.delete(this.sessionId);
    }
  }
  buildSharePageUrl(template) {
    const baseUrl = "https://share.reclaimprotocol.org/verify";
    if (this.customSharePageUrl) {
      return `${this.customSharePageUrl}/?template=${template}`;
    }
    return `${baseUrl}/?template=${template}`;
  }
  /**
   * Exports the Reclaim proof verification request as a JSON string
   *
   * This serialized format can be sent to the frontend to recreate this request using
   * ReclaimProofRequest.fromJsonString() or any InApp SDK's startVerificationFromJson()
   * method to initiate the verification journey.
   *
   * @returns JSON string representation of the proof request. Note: The JSON includes both `timestamp` and `timeStamp` (deprecated) for backward compatibility.
   *
   * @example
   * ```typescript
   * const jsonString = proofRequest.toJsonString();
   * // Send to frontend or store for later use
   * // Can be reconstructed with: ReclaimProofRequest.fromJsonString(jsonString)
   * ```
   */
  toJsonString() {
    var _a;
    return JSON.stringify({
      applicationId: this.applicationId,
      providerId: this.providerId,
      sessionId: this.sessionId,
      context: this.context,
      appCallbackUrl: this.appCallbackUrl,
      claimCreationType: this.claimCreationType,
      parameters: this.parameters,
      signature: this.signature,
      redirectUrl: this.redirectUrl,
      redirectUrlOptions: this.redirectUrlOptions,
      cancelCallbackUrl: this.cancelCallbackUrl,
      cancelRedirectUrl: this.cancelRedirectUrl,
      cancelRedirectUrlOptions: this.cancelRedirectUrlOptions,
      timestamp: this.timeStamp,
      // New field with correct spelling
      timeStamp: this.timeStamp,
      // @deprecated: Remove in future versions
      options: this.options,
      sdkVersion: this.sdkVersion,
      jsonProofResponse: this.jsonProofResponse,
      resolvedProviderVersion: (_a = this.resolvedProviderVersion) != null ? _a : "",
      modalOptions: this.modalOptions ? {
        title: this.modalOptions.title,
        description: this.modalOptions.description,
        extensionUrl: this.modalOptions.extensionUrl,
        darkTheme: this.modalOptions.darkTheme,
        modalPopupTimer: this.modalOptions.modalPopupTimer,
        showExtensionInstallButton: this.modalOptions.showExtensionInstallButton
        // onClose is intentionally excluded as functions cannot be serialized
      } : void 0
    });
  }
  /**
   * Generates and returns the request URL for proof verification
   *
   * This URL can be shared with users to initiate the proof generation process.
   * The URL format varies based on device type:
   * - Mobile iOS: Returns App Clip URL (if useAppClip is enabled)
   * - Mobile Android: Returns Instant App URL (if useAppClip is enabled)
   * - Desktop/Other: Returns standard verification URL
   *
   * @param launchOptions - Optional launch configuration to override default behavior
   * @returns Promise<string> - The generated request URL
   * @throws {SignatureNotFoundError} When signature is not set
   *
   * @example
   * ```typescript
   * const requestUrl = await proofRequest.getRequestUrl();
   * // Share this URL with users or display as QR code
   * ```
   */
  getRequestUrl(launchOptions) {
    return __async(this, null, function* () {
      var _a, _b, _c;
      const options = launchOptions || ((_a = this.options) == null ? void 0 : _a.launchOptions) || {};
      logger7.info("Creating Request Url");
      if (!this.signature) {
        throw new SignatureNotFoundError("Signature is not set.");
      }
      try {
        const templateData = this.getTemplateData();
        yield updateSession(this.sessionId, "SESSION_STARTED" /* SESSION_STARTED */);
        const deviceType = getDeviceType();
        if (((_b = this.options) == null ? void 0 : _b.useAppClip) && deviceType === "mobile" /* MOBILE */) {
          let template = encodeURIComponent(JSON.stringify(templateData));
          template = replaceAll(template, "(", "%28");
          template = replaceAll(template, ")", "%29");
          const isIos = getMobileDeviceType() === "ios" /* IOS */;
          if (!isIos) {
            let instantAppUrl = this.buildSharePageUrl(template);
            const isDeferredDeeplinksFlowEnabled = (_c = options.canUseDeferredDeepLinksFlow) != null ? _c : false;
            if (isDeferredDeeplinksFlowEnabled) {
              instantAppUrl = instantAppUrl.replace("/verifier", "/link");
            }
            logger7.info("Instant App Url created successfully: " + instantAppUrl);
            return instantAppUrl;
          } else {
            const appClipUrl = this.customAppClipUrl ? `${this.customAppClipUrl}&template=${template}` : `https://appclip.apple.com/id?p=org.reclaimprotocol.app.clip&template=${template}`;
            logger7.info("App Clip Url created successfully: " + appClipUrl);
            return appClipUrl;
          }
        } else {
          const link = yield createLinkWithTemplateData(templateData, this.customSharePageUrl);
          logger7.info("Request Url created successfully: " + link);
          return link;
        }
      } catch (error) {
        logger7.info("Error creating Request Url:", error);
        throw error;
      }
    });
  }
  /**
   * Triggers the appropriate Reclaim verification flow based on device type and configuration
   *
   * This method automatically detects the device type and initiates the optimal verification flow:
   * - Desktop with browser extension: Triggers extension flow
   * - Desktop without extension: Shows QR code modal
   * - Mobile Android: Redirects to Instant App
   * - Mobile iOS: Redirects to App Clip
   *
   * @param launchOptions - Optional launch configuration to override default behavior
   * @returns Promise<void>
   * @throws {SignatureNotFoundError} When signature is not set
   *
   * @example
   * ```typescript
   * await proofRequest.triggerReclaimFlow();
   * // The appropriate verification method will be triggered automatically
   * ```
   */
  triggerReclaimFlow(launchOptions) {
    return __async(this, null, function* () {
      var _a, _b;
      const options = launchOptions || ((_a = this.options) == null ? void 0 : _a.launchOptions) || {};
      if (!this.signature) {
        throw new SignatureNotFoundError("Signature is not set.");
      }
      try {
        const templateData = this.getTemplateData();
        this.templateData = templateData;
        logger7.info("Triggering Reclaim flow");
        const deviceType = getDeviceType();
        updateSession(this.sessionId, "SESSION_STARTED" /* SESSION_STARTED */);
        if (deviceType === "desktop" /* DESKTOP */) {
          const extensionAvailable = yield this.isBrowserExtensionAvailable();
          if (((_b = this.options) == null ? void 0 : _b.useBrowserExtension) && extensionAvailable) {
            logger7.info("Triggering browser extension flow");
            this.triggerBrowserExtensionFlow();
            return;
          } else {
            logger7.info("Browser extension not available, showing QR code modal");
            yield this.showQRCodeModal();
          }
        } else if (deviceType === "mobile" /* MOBILE */) {
          const mobileDeviceType = getMobileDeviceType();
          if (mobileDeviceType === "android" /* ANDROID */) {
            logger7.info("Redirecting to Android instant app");
            yield this.redirectToInstantApp(options);
          } else if (mobileDeviceType === "ios" /* IOS */) {
            logger7.info("Redirecting to iOS app clip");
            this.redirectToAppClip();
          }
        }
      } catch (error) {
        logger7.info("Error triggering Reclaim flow:", error);
        throw error;
      }
    });
  }
  /**
   * Checks if the Reclaim browser extension is installed and available
   *
   * This method attempts to communicate with the browser extension to verify its availability.
   * It uses a timeout mechanism to quickly determine if the extension responds.
   *
   * @param timeout - Timeout in milliseconds to wait for extension response. Defaults to 200ms
   * @returns Promise<boolean> - True if extension is available, false otherwise
   *
   * @example
   * ```typescript
   * const hasExtension = await proofRequest.isBrowserExtensionAvailable();
   * if (hasExtension) {
   *   console.log('Browser extension is installed');
   * }
   * ```
   */
  isBrowserExtensionAvailable(timeout = 200) {
    return __async(this, null, function* () {
      try {
        return new Promise((resolve) => {
          const messageId = `reclaim-check-${Date.now()}`;
          const timeoutId = setTimeout(() => {
            window.removeEventListener("message", messageListener);
            resolve(false);
          }, timeout);
          const messageListener = (event) => {
            var _a, _b;
            if (((_a = event.data) == null ? void 0 : _a.action) === RECLAIM_EXTENSION_ACTIONS.EXTENSION_RESPONSE && ((_b = event.data) == null ? void 0 : _b.messageId) === messageId) {
              clearTimeout(timeoutId);
              window.removeEventListener("message", messageListener);
              resolve(!!event.data.installed);
            }
          };
          window.addEventListener("message", messageListener);
          const message = {
            action: RECLAIM_EXTENSION_ACTIONS.CHECK_EXTENSION,
            extensionID: this.extensionID,
            messageId
          };
          window.postMessage(message, "*");
        });
      } catch (error) {
        logger7.info("Error checking Reclaim extension installed:", error);
        return false;
      }
    });
  }
  triggerBrowserExtensionFlow() {
    const message = {
      action: RECLAIM_EXTENSION_ACTIONS.START_VERIFICATION,
      messageId: this.sessionId,
      data: this.templateData,
      extensionID: this.extensionID
    };
    window.postMessage(message, "*");
    logger7.info("Browser extension flow triggered");
  }
  showQRCodeModal() {
    return __async(this, null, function* () {
      try {
        const requestUrl = yield createLinkWithTemplateData(this.templateData, this.customSharePageUrl);
        this.modal = new QRCodeModal(this.modalOptions);
        yield this.modal.show(requestUrl);
      } catch (error) {
        logger7.info("Error showing QR code modal:", error);
        throw error;
      }
    });
  }
  redirectToInstantApp(options) {
    return __async(this, null, function* () {
      var _a;
      try {
        let template = encodeURIComponent(JSON.stringify(this.templateData));
        template = replaceAll(template, "(", "%28");
        template = replaceAll(template, ")", "%29");
        let instantAppUrl = this.buildSharePageUrl(template);
        logger7.info("Redirecting to Android instant app: " + instantAppUrl);
        const isDeferredDeeplinksFlowEnabled = (_a = options.canUseDeferredDeepLinksFlow) != null ? _a : false;
        if (isDeferredDeeplinksFlowEnabled) {
          instantAppUrl = instantAppUrl.replace("/verifier", "/link");
          const deepLink = `intent://details?id=org.reclaimprotocol.app&url=${encodeURIComponent(
            instantAppUrl
          )}&template=${template}#Intent;scheme=market;action=android.intent.action.VIEW;package=com.android.vending;end;`;
          try {
            const requestUrl = instantAppUrl;
            let appInstalled = false;
            let timeoutId;
            const iframe = document.createElement("iframe");
            iframe.style.display = "none";
            iframe.style.width = "1px";
            iframe.style.height = "1px";
            document.body.appendChild(iframe);
            const cleanup = () => {
              if (iframe.parentNode) {
                document.body.removeChild(iframe);
              }
              if (timeoutId) {
                clearTimeout(timeoutId);
              }
            };
            const onVisibilityChange = () => {
              if (document.hidden) {
                appInstalled = true;
                cleanup();
                window.location.href = deepLink;
              }
            };
            document.addEventListener("visibilitychange", onVisibilityChange, { once: true });
            iframe.src = deepLink.replace("intent:", "reclaimverifier:");
            timeoutId = setTimeout(() => {
              document.removeEventListener("visibilitychange", onVisibilityChange);
              cleanup();
              if (!appInstalled) {
                window.navigator.clipboard.writeText(requestUrl).catch(() => {
                  console.error("We can't access the clipboard. Please copy this link and open Reclaim Verifier app.");
                });
                window.location.href = deepLink;
              }
            }, 1500);
          } catch (e) {
            window.location.href = instantAppUrl;
          }
          return;
        }
        window.location.href = instantAppUrl;
      } catch (error) {
        logger7.info("Error redirecting to instant app:", error);
        throw error;
      }
    });
  }
  redirectToAppClip() {
    try {
      let template = encodeURIComponent(JSON.stringify(this.templateData));
      template = replaceAll(template, "(", "%28");
      template = replaceAll(template, ")", "%29");
      const appClipUrl = this.customAppClipUrl ? `${this.customAppClipUrl}&template=${template}` : `https://appclip.apple.com/id?p=org.reclaimprotocol.app.clip&template=${template}`;
      logger7.info("Redirecting to iOS app clip: " + appClipUrl);
      const verifierUrl = `https://share.reclaimprotocol.org/verifier/?template=${template}`;
      window.location.href = appClipUrl;
      setTimeout(() => {
        window.location.href = verifierUrl;
      }, 5 * 1e3);
    } catch (error) {
      logger7.info("Error redirecting to app clip:", error);
      throw error;
    }
  }
  /**
   * Starts the proof request session and monitors for proof submission
   *
   * This method begins polling the session status to detect when
   * a proof has been generated and submitted. It handles both default Reclaim callbacks
   * and custom callback URLs.
   *
   * For default callbacks: Verifies proofs automatically and passes them to onSuccess
   * For custom callbacks: Monitors submission status and notifies via onSuccess when complete.
   * In the custom-callback flow (where the SDK submits a proof to a provided callback URL),
   * onSuccess may be invoked with an empty array (onSuccess([])) when no proof is available
   * (this happens when a callback is set using setAppCallbackUrl where proof is sent to callback instead of reclaim backend).
   *
   * Please refer to the OnSuccess type signature ((proof?: Proof | Proof[]) => void)
   * and the startSession function source for more details.
   *
   * > [!TIP]
   * > **Best Practice:** When using `setAppCallbackUrl` and/or `setCancelCallbackUrl`, your backend receives the proof or cancellation details directly. We recommend your backend notifies the frontend (e.g. via WebSockets, SSE, or polling) to stop the verification process and handle the appropriate success/failure action. Do not rely completely on `startSession` callbacks on the frontend when using these backend callbacks.
   *
   * @param onSuccess - Callback function invoked when proof is successfully submitted
   * @param onError - Callback function invoked when an error occurs during the session
   * @returns Promise<void>
   * @throws {SessionNotStartedError} When session ID is not defined
   * @throws {ProofNotVerifiedError} When proof verification fails (default callback only)
   * @throws {ProofSubmissionFailedError} When proof submission fails (custom callback only)
   * @throws {ProviderFailedError} When proof generation fails with timeout
   *
   * @example
   * ```typescript
   * await proofRequest.startSession({
   *   onSuccess: (proof) => {
   *     console.log('Proof received:', proof);
   *   },
   *   onError: (error) => {
   *     console.error('Error:', error);
   *   }
   * });
   * ```
   */
  startSession(_0) {
    return __async(this, arguments, function* ({ onSuccess, onError }) {
      if (!this.sessionId) {
        const message = "Session can't be started due to undefined value of sessionId";
        logger7.info(message);
        throw new SessionNotStartedError(message);
      }
      logger7.info("Starting session");
      const sessionUpdatePollingInterval = 3 * 1e3;
      const interval = setInterval(() => __async(this, null, function* () {
        var _a, _b, _c, _d;
        try {
          const statusUrlResponse = yield fetchStatusUrl(this.sessionId);
          if (!statusUrlResponse.session) return;
          if (statusUrlResponse.session.statusV2 !== "PROOF_GENERATION_FAILED" /* PROOF_GENERATION_FAILED */) {
            this.lastFailureTime = void 0;
          }
          if (statusUrlResponse.session.statusV2 === "PROOF_GENERATION_FAILED" /* PROOF_GENERATION_FAILED */) {
            const currentTime = Date.now();
            if (!this.lastFailureTime) {
              this.lastFailureTime = currentTime;
            } else if (currentTime - this.lastFailureTime >= this.FAILURE_TIMEOUT) {
              throw new ProviderFailedError("Proof generation failed - timeout reached");
            }
            return;
          }
          if (statusUrlResponse.session.statusV2 === "ERROR_SUBMISSION_FAILED" /* ERROR_SUBMISSION_FAILED */ || statusUrlResponse.session.statusV2 === "ERROR_SUBMITTED" /* ERROR_SUBMITTED */) {
            throw new ErrorDuringVerificationError();
          }
          const isDefaultCallbackUrl = this.getAppCallbackUrl() === `${constants.DEFAULT_RECLAIM_CALLBACK_URL}${this.sessionId}`;
          if (isDefaultCallbackUrl) {
            if (statusUrlResponse.session.proofs && statusUrlResponse.session.proofs.length > 0) {
              const proofs = statusUrlResponse.session.proofs;
              if (this.claimCreationType === "createClaim" /* STANDALONE */) {
                const verified = yield verifyProof(proofs, (_a = this.options) == null ? void 0 : _a.acceptAiProviders);
                if (!verified) {
                  logger7.info(`Proofs not verified: ${JSON.stringify(proofs)}`);
                  throw new ProofNotVerifiedError();
                }
              }
              if (proofs.length === 1) {
                onSuccess(proofs[0]);
              } else {
                onSuccess(proofs);
              }
              this.clearInterval();
              (_b = this.modal) == null ? void 0 : _b.close();
            }
          } else {
            if (statusUrlResponse.session.statusV2 === "PROOF_SUBMISSION_FAILED" /* PROOF_SUBMISSION_FAILED */) {
              throw new ProofSubmissionFailedError();
            }
            if (statusUrlResponse.session.statusV2 === "PROOF_SUBMITTED" /* PROOF_SUBMITTED */ || statusUrlResponse.session.statusV2 === "AI_PROOF_SUBMITTED" /* AI_PROOF_SUBMITTED */) {
              if (onSuccess) {
                onSuccess([]);
              }
              this.clearInterval();
              (_c = this.modal) == null ? void 0 : _c.close();
            }
          }
        } catch (e) {
          if (onError) {
            onError(e);
          }
          this.clearInterval();
          (_d = this.modal) == null ? void 0 : _d.close();
        }
      }), sessionUpdatePollingInterval);
      this.intervals.set(this.sessionId, interval);
      scheduleIntervalEndingTask(this.sessionId, this.intervals, onError);
    });
  }
  /**
   * Closes the QR code modal if it is currently open
   *
   * This method can be called to programmatically close the modal, for example,
   * when implementing custom UI behavior or cleanup logic.
   *
   * @example
   * ```typescript
   * // Close modal after some condition
   * proofRequest.closeModal();
   * ```
   */
  closeModal() {
    if (this.modal) {
      this.modal.close();
      logger7.info("Modal closed by user");
    }
  }
  /**
   * Returns whether proofs will be submitted as JSON format
   *
   * @returns boolean - True if proofs are sent as application/json, false for application/x-www-form-urlencoded
   *
   * @example
   * ```typescript
   * const isJson = proofRequest.getJsonProofResponse();
   * console.log('JSON format:', isJson);
   * ```
   */
  getJsonProofResponse() {
    return this.jsonProofResponse;
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ReclaimProofRequest,
  assertValidProof,
  clearDeviceCache,
  getDeviceType,
  getMobileDeviceType,
  isDesktopDevice,
  isMobileDevice,
  transformForOnchain,
  verifyProof
});
